source/extensions/filters/http/jwt_authn/authenticator.cc

```diff
@@ -0,0 +1,266 @@
+  // The on_done function.
+  Authenticator::Callbacks* callback_{};
+
+  // The pending uri_, only used for logging.
+  std::string uri_;
+  // The pending remote request so it can be canceled.
+  Http::AsyncClient::Request* request_{};
+};
+
+void AuthenticatorImpl::sanitizePayloadHeaders(Http::HeaderMap& headers) const {
+  for (const auto& rule : config_->getProtoConfig().rules()) {
+    if (!rule.forward_payload_header().empty()) {
+      headers.remove(Http::LowerCaseString(rule.forward_payload_header()));
+    }
+  }
+}
+
+void AuthenticatorImpl::verify(Http::HeaderMap& headers, Authenticator::Callbacks* callback) {
+  headers_ = &headers;
+  callback_ = callback;
+
+  ENVOY_LOG(debug, "Jwt authentication starts");
+  auto tokens = config_->getExtractor().extract(headers);
+  if (tokens.empty()) {
+    if (okToBypass()) {
+      doneWithStatus(Status::Ok);
+    } else {
+      doneWithStatus(Status::JwtMissed);
+    }
+    return;
+  }
+
+  // TODO(qiwzhang), add supports for multiple tokens.
+  // Only process the first token for now.
+  token_.swap(tokens[0]);
+
+  const Status status = jwt_.parseFromString(token_->token());
+  if (status != Status::Ok) {
+    doneWithStatus(status);
+    return;
+  }
+
+  // Check if token is extracted from the location specified by the issuer.
+  if (!token_->isIssuerSpecified(jwt_.iss_)) {
+    ENVOY_LOG(debug, "Jwt for issuer {} is not extracted from the specified locations", jwt_.iss_);
+    doneWithStatus(Status::JwtUnknownIssuer);
+    return;
+  }
+
+  // Check "exp" claim.
+  const auto unix_timestamp = std::chrono::duration_cast<std::chrono::seconds>(
+                                  std::chrono::system_clock::now().time_since_epoch())
+                                  .count();
+  if (jwt_.exp_ < unix_timestamp) {
+    doneWithStatus(Status::JwtExpired);
+    return;
+  }
+
+  // Check the issuer is configured or not.
+  jwks_data_ = config_->getCache().getJwksCache().findByIssuer(jwt_.iss_);
+  if (jwks_data_ == nullptr) {
+    doneWithStatus(Status::JwtUnknownIssuer);
+    return;
+  }
+
+  // Check if audience is allowed
+  if (!jwks_data_->areAudiencesAllowed(jwt_.audiences_)) {
+    doneWithStatus(Status::JwtAudienceNotAllowed);
+    return;
+  }
+
+  if (jwks_data_->getJwksObj() != nullptr && !jwks_data_->isExpired()) {
+    verifyKey();
+    return;
+  }
+
+  // TODO(qiwzhang): potential optimization.
+  // If request 1 triggers a remote jwks fetching, but is not yet replied when the request 2
+  // of using the same jwks comes. The request 2 will trigger another remote fetching for the
+  // jwks. This can be optimized; the same remote jwks fetching can be shared by two requrests.
+  fetchRemoteJwks();
+}
+
+void AuthenticatorImpl::fetchRemoteJwks() {
+  uri_ = jwks_data_->getJwtRule().remote_jwks().http_uri().uri();
+  std::string host, path;
+  Http::Utility::extractHostPathFromUri(uri_, host, path);
+
+  Http::MessagePtr message(new Http::RequestMessageImpl());
```

Comment:
can we make this part an utility function for `HttpUri`, similar to `Grpc::Common::prepareHeaders`.

Reply:
Done